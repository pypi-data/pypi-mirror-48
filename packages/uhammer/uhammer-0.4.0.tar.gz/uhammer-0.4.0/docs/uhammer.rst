uhammer package
===============

All listed classes and functions are available by importing
from `uhammer`.

uhammer.lnprob\_adapters module
-------------------------------

.. automodule:: uhammer.lnprob_adapters
    :members:
    :undoc-members:
    :show-inheritance:

uhammer.parameters module
-------------------------

.. automodule:: uhammer.parameters
    :members:
    :undoc-members:
    :show-inheritance:

uhammer.persisting module
-------------------------

.. automodule:: uhammer.persisting
    :members:
    :undoc-members:
    :show-inheritance:

uhammer.sampler module
----------------------

.. automodule:: uhammer.sampler
    :members:
    :undoc-members:
    :show-inheritance:


Module contents
---------------

.. automodule:: uhammer
    :members:
    :undoc-members:
    :show-inheritance:
