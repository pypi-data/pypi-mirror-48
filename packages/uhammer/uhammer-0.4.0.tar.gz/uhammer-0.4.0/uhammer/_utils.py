#! /usr/bin/env python
# encoding: utf-8
from __future__ import absolute_import, division, print_function

import functools
import io
import os
import random
import sys
import threading
import time
from contextlib import contextmanager

try:
    from shutil import get_terminal_size
except ImportError:  # pragma: no cover
    from backports.shutil_get_terminal_size import get_terminal_size  # pragma: no cover


is_py_3 = sys.version_info.major == 3


def _console_dimensions():

    try:
        columns, rows = get_terminal_size((80, 25))
    except OSError:
        return 80, 25
    return columns, rows  # pragma: no cover


def _format_time(t):
    if t is None:
        return "??:??:??.?"
    minutes, seconds = divmod(t, 60)
    hours, minutes = divmod(minutes, 60)
    return "{:02.0f}:{:02.0f}:{:04.1f}".format(hours, minutes, seconds)


def _format_time_part(t_passed, t_left):
    time_str_passed = "passed: " + _format_time(t_passed)
    time_str_left = "left: " + _format_time(t_left)
    return time_str_passed + " " + time_str_left + " - "


class _Formatter(object):
    def __init__(self, bar_size, n_steps, started):
        self.bar_size = bar_size
        self.n_steps = n_steps
        self.started = started
        self.time_left = 0

    def format_progress_line(self, symbol, i, tl=None):
        n = self.bar_size * i // self.n_steps
        time_passed = time.time() - self.started

        progress_symbol = "∣"

        if tl is None:
            time_left_estimate = time_passed / (i + 1) * (self.n_steps - i - 1)

            # exponential smoothing, avoid "flattering" for fast sampling:
            self.time_left = 0.8 * self.time_left + 0.2 * time_left_estimate
        else:
            self.time_left = tl
        time_str = _format_time_part(time_passed, self.time_left)
        return "\r{} {}[{}{}]".format(
            symbol, time_str, progress_symbol * n, " " * (self.bar_size - n)
        )


def _write(what):
    # write to systems stdout directly as sys.stdout might be redirected due to
    # capturing
    if is_py_3:
        os.write(1, what.encode("utf-8"))
    else:
        os.write(1, what)  # pragma: no cover


class _UpdateThread(threading.Thread):
    def __init__(self, symbols, formatter, done_symbol, dt=0.07):
        threading.Thread.__init__(self)
        self.symbols = symbols
        self.formatter = formatter
        self.done_symbol = done_symbol
        self.dt = dt

        self.progress_counter = 0

    def run(self):
        self.iter_running = True
        symbol_idx = 0
        try:
            while self.iter_running:
                time.sleep(self.dt)
                symbol = self.symbols[symbol_idx]
                symbol_idx = (symbol_idx + 1) % len(self.symbols)
                progress_line = self.formatter.format_progress_line(
                    symbol, self.progress_counter
                )
                _write(progress_line)
            progress_line = self.formatter.format_progress_line(
                self.done_symbol, self.progress_counter, 0
            )
            _write(progress_line)
            _write("\n")
        except KeyboardInterrupt:
            pass


def with_progressbar(n_steps, iterable, width=None, delta_t=0.05):

    if width is None:
        width, __ = _console_dimensions()

    symbols = random.choice(
        [
            "◰ ◳ ◲ ◱",
            "▁▂▃▄▅▆▇███▇▆▅▄▃▁▁",
            "▏▎▍▌▋▊███▊▋▌▍▎▏",
            "▏▎▍▌▋▊▉███▏",
            "▖▘▝▗",
            "◐ ◓ ◑ ◒ ",
            "◴ ◷ ◶ ◵",
            "⣾⣽⣻⢿⡿⣟⣯⣷",
            "⠁⠂⠄⡀⢀⠠⠐⠈",
            "◢ ◣ ◤ ◥ ",
            ".oOo",
            "⎺⎻⎼⎽⎼⎻⎺",
            "⎽⎼⎻⎺⎺⎽",
        ]
    ).replace(" ", "")

    done_symbol = "✗"

    time_str = _format_time_part(0, None)
    bar_size = min(n_steps, width - 6 - len(time_str))

    started = time.time()
    formatter = _Formatter(bar_size, n_steps, started)

    thread = _UpdateThread(symbols, formatter, done_symbol)
    thread.start()

    try:
        for result in iterable:
            thread.progress_counter += 1
            yield result
            if not thread.is_alive():
                # usually KeyboardInterrupt happened in update thread
                break
    except KeyboardInterrupt:  # pragma: no cover
        thread.iter_running = False  # pragma: no cover
        thread.join()  # pragma: no cover
        raise  # pragma: no cover

    finally:
        # stop thread:
        thread.iter_running = False
        thread.join()


def with_simple_progress_report(n_steps, iterable, minimum_time_interval=30):
    started = time.time()
    time_left = None

    last_report = started

    for i, result in enumerate(iterable):

        if time.time() > last_report + minimum_time_interval:
            time_left = _print_progress(started, time_left, i, n_steps)
            last_report = time.time()

        yield result


def _print_progress(started, time_left, i, n_steps):
    time_passed = time.time() - started

    time_left_estimate = time_passed / (i + 1) * (n_steps - i - 1)
    if time_left is not None:
        time_left = 0.8 * time_left + 0.2 * time_left_estimate
    else:
        time_left = time_left_estimate

    tp = _format_time(time_passed)
    tl = _format_time(time_left)

    print(
        "time passed: {}  estimated time left: {}  computed {} / {} samples".format(
            tp, tl, i + 1, n_steps
        )
    )
    return time_left


def is_float(x):
    try:
        float(x)
    except ValueError:
        return False
    else:
        return True


def check_range(range_):
    if not isinstance(range_, (list, tuple)) or len(range_) != 2:
        raise ValueError(
            "range_ must be list or tuple of length 2, got {!r}".format(range_)
        )

    min_, max_ = range_

    if not is_float(min_) or not is_float(max_):
        raise ValueError(
            "range_ must be a tuple of 2 float values, got {!r}".format(range_)
        )
    if not min_ < max_:
        raise ValueError(
            "min value {} is not smaller than max value {}".format(min_, max_)
        )


def is_writable(path):

    if os.path.exists(path):
        return os.access(path, os.W_OK)
    try:
        open(path, "w").close()
    except IOError as e:
        print("ERROR", e)
        return False
    try:
        os.unlink(path)
    except FileNotFoundError as e:  # possible race condition when ran in parallel
        print("ERROR", e)
        pass
    return True


class Tee(io.TextIOBase):
    def __init__(self, file_stream, console_stream):
        self.file_stream = file_stream
        self.console_stream = console_stream

    def write(self, *args):
        if self.file_stream is not None:
            self.file_stream.write(*args)
        if self.console_stream is not None:
            self.console_stream.write(*args)

    def close(self):
        pass


@contextmanager
def capture(path, show_output):

    if path is not None:
        fh = open(path, "a")
    else:
        fh = None

    stdout_stream = Tee(fh, sys.stdout if show_output else None)
    stderr_stream = Tee(fh, sys.stderr if show_output else None)

    old_stdout = sys.stdout
    old_stderr = sys.stderr
    try:
        sys.stdout = stdout_stream
        sys.stderr = stderr_stream
        yield

    finally:
        sys.stdout = old_stdout
        sys.stderr = old_stderr
        if fh is not None:
            fh.close()


def os_write_works():
    try:
        os.write(1, b"")
    except IOError:
        return False
    else:
        return True


# from https://stackoverflow.com/questions/6796492/temporarily-redirect-stdout-stderr:


@contextmanager
def stdchannel_redirected(stdchannel, dest_filename):
    """
    A context manager to temporarily redirect stdout or stderr

    e.g.:

    with stdchannel_redirected(sys.stderr, os.devnull):
        ...
    """

    oldstdchannel = None
    dest_file = None
    dest_file = open(dest_filename, "w")
    try:
        oldstdchannel = os.dup(stdchannel.fileno())
        os.dup2(dest_file.fileno(), stdchannel.fileno())
        yield
    except IOError:
        # sth with stream handling went wrong, happens e.g. in Jupyter notebooks.
        # In this case we don't redirect:
        yield
    finally:
        if oldstdchannel is not None:
            os.dup2(oldstdchannel, stdchannel.fileno())
        if dest_file is not None:
            dest_file.close()


mute_stderr = functools.partial(stdchannel_redirected, sys.stderr, os.devnull)
