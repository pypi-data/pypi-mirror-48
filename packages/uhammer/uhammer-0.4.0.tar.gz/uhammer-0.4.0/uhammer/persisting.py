#! /usr/bin/env python
# encoding: utf-8
from __future__ import absolute_import, division, print_function

import zlib

import dill
import emcee

emcee_version = int(emcee.__version__.split(".")[0])

# Copyright Â© 2018 Uwe Schmitt <uwe.schmitt@id.ethz.ch>


class _SamplerPersister(object):
    def __init__(self, check, path, persist_on_exceptions):
        self.check = check
        self.path = path
        self.persist_on_exceptions = persist_on_exceptions
        self._iter_track = []

    def __call__(self, sampler, i, n_iter, exception):
        do_persist = self.check(i, n_iter)
        do_persist = do_persist or (
            exception is not None and self.persist_on_exceptions
        )
        if do_persist:
            with open(self.path, "wb") as fh:
                fh.write(zlib.compress(dill.dumps(sampler)))
            self._iter_track.append(i)
        if exception:
            raise exception
        return do_persist

    @staticmethod
    def restore_sampler(file_path):
        with open(file_path, "rb") as fh:
            return dill.loads(zlib.decompress(fh.read()))


def load_samples(path):
    """loads samples for persisted sampler file"""
    sampler = _SamplerPersister.restore_sampler(path)
    if emcee_version == 3:
        return sampler.flatchain
    n_samples = sampler.iterations * sampler.n_walkers
    return sampler.flatchain[:n_samples]


def persist_every_n_iterations(n, path):
    """Return value of this function can be passed as argument `persist` to ``sample``
    function to store state during sampling. This allows later continuation of sampling,
    e.g. after crashes.

    This persister saves state every ``n`` iterations and in case of failure.

    :param n:  persist after every ``n`` iterations of the sampler
    :param path: file path to store the relevant information.
    """

    def check(i, n_iter):
        return i > 0 and i % n == 0

    return _SamplerPersister(check, path, True)


def persist_final(path):
    """Return value of this function can be passed as argument `persist` to ``sample``
    function to store state during sampling. This allows later continuation of sampling,
    e.g. after crashes.

    This persister only stores the information when sampling finishes and in case
    of failure.

    :param path: file path to store the relevant information.
    """

    def check(i, n_iter):
        return i == n_iter

    return _SamplerPersister(check, path, True)


def persist_on_error(path):
    """Return value of this function can be passed as argument `persist` to ``sample``
    function to store state during sampling. This allows later continuation of sampling,
    e.g. after crashes.

    This persister only stores the information only in case of failure.

    :param path: file path to store the relevant information.
    """

    def check(i, n_iter):
        return False

    return _SamplerPersister(check, path, True)


def dont_persist():
    def check(i, n_iter):
        return False

    return _SamplerPersister(check, None, False)
