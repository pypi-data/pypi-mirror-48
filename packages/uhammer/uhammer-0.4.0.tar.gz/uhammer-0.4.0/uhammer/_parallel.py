import ipaddress
import os
import socket
import sys
import warnings
from multiprocessing import Queue, cpu_count, current_process

from schwimmbad import MultiPool as _MultiPool

from ._utils import is_writable, mute_stderr

try:
    from mpipool import Pool as _MPIPool
except ImportError:

    class _MPIPool(object):
        def __getattr__(self, name):
            raise ImportError("you must run 'pip install mpipool' first.")


if sys.version_info.major == 2:
    ModuleNotFoundError = None
    unicode = str


class UMultiPool(_MultiPool):

    def __str__(self):
        return "MultiPool"


class MPIPool(_MPIPool):

    def __str__(self):
        return "MPIPool"


def runs_on_euler_node():
    # ip ranges according to https://scicomp.ethz.ch/wiki/Cluster_IP_ranges#Euler
    try:
        ip4_text = ipaddress.IPv4Address(socket.gethostbyname(socket.gethostname()))
        return any(
            ip4_text in ipaddress.IPv4Network(network)
            for network in ("10.205.0.0/19", "10.205.96.0/19")
        )
    except Exception:
        ip4_text = ipaddress.IPv4Address(
            unicode(socket.gethostbyname(socket.gethostname()))
        )
        return any(
            ip4_text in ipaddress.IPv4Network(network)
            for network in (unicode("10.205.0.0/19"), unicode("10.205.96.0/19"))
        )


def runs_on_single_euler_node():
    resources = os.environ.get("LSB_EFFECTIVE_RSRCREQ", "")
    return "span[hosts=1]" in resources


def runs_with_mpi():
    return any(
        n.startswith(prefix)
        for n in os.environ.keys()
        for prefix in ("MPIEXEC_", "OMPI_COMM_WORLD_")
    )


def mpi_size():
    if runs_with_mpi():
        from mpi4py import MPI

        return MPI.COMM_WORLD.size
    return None


def _inject_rank_information(path, rank):
    if "{worker_id" in path:
        path = path.format(worker_id=rank)
    return path


def check_and_fix_path(path, rank):

    if rank is not None:
        path = _inject_rank_information(path, rank)

    path = os.path.abspath(path)
    folder = os.path.dirname(path)

    if folder and not os.path.exists(folder):
        os.makedirs(folder)

    if os.path.exists(path):
        raise IOError("{} already exists".format(path))

    if not is_writable(path):
        raise IOError(
            "can not write to {}. please check if folder {} is writable".format(
                path, folder
            )
        )
    return path


def get_rank():
    if runs_with_mpi():
        from mpi4py import MPI

        return MPI.COMM_WORLD.rank

    return getattr(current_process(), "rank", 0)  # main process has no rank set


def core_count():
    lsb_cores = os.environ.get("LSB_MAX_NUM_PROCESSORS")
    if lsb_cores is None:
        return cpu_count()
    return int(lsb_cores)


def mpi_tasks():
    return _mpi_query("size", default=1)


def _mpi_query(attribute, default=None):

    if not runs_with_mpi():
        return None
    try:
        from mpi4py import MPI
    except ImportError as e:
        if ModuleNotFoundError is not None and isinstance(e, ModuleNotFoundError):
            warnings.warn("mpi4py is not installed.")
            return default
        warnings.warn("mpi4py is installed, but I can not import it: {}.".format(e))
        return default
    return getattr(MPI.COMM_WORLD, attribute)  # pragma: no cover


def mpi_rank():
    return _mpi_query("rank", default=1)


def check_if_walkers_number_of_cores_fit(n_walkers, n_workers):

    if n_workers == 1:
        warnings.warn("only one worker available. switch to serial mode.")
        return False

    if n_walkers < n_workers:
        msg = (
            "the number of walkers {n_walkers} is less than the number of workers "
            "{n_workers}.".format(n_walkers=n_walkers, n_workers=n_workers)
        )
        warnings.warn(msg)

    else:
        if n_walkers % n_workers:
            msg = (
                "uhammer runs with {n_workers} workers for {n_walkers} walkers. "
                "resource usage is best if the number walkers is a multiple of number "
                "of workers.".format(n_walkers=n_walkers, n_workers=n_workers)
            )
            warnings.warn(msg)
    return True


def get_pool_for(n_walkers):

    pool = _get_pool(n_walkers)
    check_pool(pool, n_walkers)
    return pool


def check_parallel(parallel):
    if not parallel and runs_with_mpi():
        warnings.warn(
            "looks like you use mpirun to run your code, but set parallel=False"
        )


def _set_rank(q):
    current_process().rank = q.get()


def _get_pool(n_walkers):

    tasks = mpi_tasks()

    if tasks is not None and tasks >= 2:  # at least one master and one worker
        return MPIPool()  # pragma: no cover

    n_workers = core_count() - 1

    # it does not make sense to use more cores than walkers
    n_cores = min(n_workers, n_walkers)

    # setup queue to assign rank attribut to all worker processes:
    q = Queue(n_cores)
    for rank in range(1, n_cores + 1):
        q.put(rank)

    with mute_stderr():
        # might cause MPI related error message, even without mpirun:
        pool = UMultiPool(n_cores, initializer=_set_rank, initargs=(q,))
        current_process().rank = 0

    return pool


def check_pool(pool, n_walkers):

    omp_threads = os.environ.get("OMP_NUM_THREADS", 1)
    try:
        omp_threads = int(omp_threads)
    except ValueError:
        raise RuntimeError(
            "OMP_NUM_THREADS has value {} which is not an integer".format(omp_threads)
        )
    if omp_threads > 1:
        warnings.warn(
            "you set OMP_NUM_THREADS to {}. this could cause very slow execution "
            "due to oversubscription. to fix this set OMP_NUM_THREADS to 1".format(
                omp_threads
            )
        )

    check_passed = check_if_walkers_number_of_cores_fit(n_walkers, pool.size)
    if not check_passed:
        return None

    if isinstance(pool, MPIPool):
        return  # pragma: no cover

    if runs_on_euler_node():

        if pool.size > 35:
            warnings.warn(
                "on euler best use -n 24 or -n 36 for scalability. "
                "larger numbers don't increase performance."
            )
        if not runs_on_single_euler_node():
            warnings.warn(
                "better run the job with '-R fullnode' to benefit "
                "from multicore parallelisation."
            )
