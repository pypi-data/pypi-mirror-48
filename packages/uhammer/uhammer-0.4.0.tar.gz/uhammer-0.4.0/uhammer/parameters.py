#! /usr/bin/env python
# encoding: utf-8
from __future__ import absolute_import, division, print_function

import warnings

import numpy as np

from ._utils import check_range


def _rand_in_range(range_):
    min_, max_ = range_
    return np.random.random() * (max_ - min_) + min_


class Parameters(object):
    """
    Collection of used parameters incl. names and ranges.
    """

    def __init__(self):
        self.limits = {}
        self.names = []

    def add(self, name, range_, start_value=None):
        """
        Adds a parameter with given `name` and value `range_` to the collection.

        :param name: Name of type ``str``.
        :param range_: Allowed range as ``tuple`` or ``list`` of two ``float``
                       values. First value must be smaller than the second.
        :param start_value: optional value to start sampling from, reduces burn-in
                            phase if estimation for parameter exists.
        """

        if start_value is not None:
            warnings.warn(
                "ignore given start_value, did not work with emcee", DeprecationWarning
            )

        check_range(range_)

        if name in self.names:
            raise ValueError("parameter named {!r} already added".format(name))

        self.limits[name] = range_
        self.names.append(name)

    def _is_in_bounds(self, name, value):
        min_, max_ = self.limits[name]
        return min_ <= value <= max_

    def _range_length(self, name):
        min_, max_ = self.limits[name]
        return max_ - min_

    def __len__(self):
        return len(self.names)

    def start_value(self, name):
        assert name in self.limits, (name, self.limits)
        return _rand_in_range(self.limits[name])
