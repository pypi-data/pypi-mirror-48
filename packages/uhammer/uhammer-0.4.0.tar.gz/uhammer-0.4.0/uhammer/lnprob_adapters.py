#! /usr/bin/env python
# encoding: utf-8
from __future__ import absolute_import, division, print_function

import numpy as np

from ._utils import check_range, is_float
from .parameters import Parameters

# Copyright Â© 2018 Uwe Schmitt <uwe.schmitt@id.ethz.ch>


def gaussian(cov):
    """Returns logarithmic probability density function for gaussian with mean 0 and
    given covariance `cov`.

    :param cov: Covariance as a single number (real case) or square invertible
                matrix (multivariate case).
    """

    class ComputeProb(object):
        def __init__(self, cov):
            self.icov = None
            self.cov = cov

        def _setup_icov(self, n):
            cov = np.atleast_1d(self.cov)
            if cov.shape == (1,):
                icov = 1.0 / cov * np.eye(n)
            else:
                icov = np.linalg.inv(cov)
            assert icov.shape == (
                n,
                n,
            ), "'cov' must be a square matrix of size {} but has shape {} x {}".format(
                n, icov.shape[0], icov.shape[1]
            )
            self.icov = icov

        def __call__(self, x):
            if self.icov is None:
                self._setup_icov(len(x))
            return -np.dot(x.T, np.dot(self.icov, x)) / 2.0

    return ComputeProb(cov)


def model(forward, probability, input_data, measured_data):
    """
    Creates a function used as `lnprob` argument for ``uhammer.sample`` function..

    :param forward: models forward computation. Takes two arguments: `p`
                    which must be an instance of ``Parameters`` and the given
                    ``input_data``.

    :param probability: A function which computes to probability of difference
                        of the computed forward model and the given
                        ``measured_data``.

    :param input_data: Input data passed as second argument to ``forward``.

    :param measured_data: Measured data. Must be compatible with the return
                          value of ``forward``.
    """

    def likelihood(p, *args):
        diff = forward(p, input_data) - measured_data
        return probability(diff)

    return likelihood


def distribution(prob_density_function, ndim, ranges):
    """
    Helper to sample from a distribution.

    Returns ``lnprob`` function and ``Parameters`` object to be used in sample.

    :param prob_density_function: Function taking vector of length ``ndim``,
                                  must return logarithm of a probability density.

    :param ndim: Dimension of input to ``prob_density_function``.

    :param ranges: Either tuple or list of tuples of numbers to restrict coordinates
                   when sampling.
    """

    tl = (tuple, list)

    assert isinstance(ranges, tl)
    assert len(ranges) > 0

    if len(ranges) == 2:
        check_range(ranges)
        min_, max_ = ranges
        if is_float(min_) and is_float(max_):
            assert min_ < max_, ranges
            ranges = [ranges] * ndim

    for range_ in ranges:
        check_range(range_)

    p = Parameters()
    for i, range_ in enumerate(ranges):
        p.add("x{i}".format(i=i), range_)

    def lnprob(p):
        return prob_density_function(np.array(p.values))

    return lnprob, p
