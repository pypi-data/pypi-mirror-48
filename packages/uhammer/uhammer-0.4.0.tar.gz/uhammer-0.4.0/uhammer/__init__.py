# encoding: utf-8
import sys

import pkg_resources

from ._parallel import runs_with_mpi as _runs_with_mpi
from .lnprob_adapters import *  # noqa
from .parameters import Parameters  # noqa
from .persisting import load_samples  # noqa
from .persisting import persist_every_n_iterations  # noqa
from .persisting import persist_final  # noqa
from .persisting import persist_on_error  # noqa
from .sampler import *  # noqa

if _runs_with_mpi():

    def handler(type_, value, tb):
        import traceback
        from mpi4py import MPI

        traceback.print_tb(tb, file=sys.stdout)
        sys.stdout.flush()

        MPI.COMM_WORLD.Abort(1)

    sys.excepthook = handler

__version__ = pkg_resources.require(__package__)[0].version
