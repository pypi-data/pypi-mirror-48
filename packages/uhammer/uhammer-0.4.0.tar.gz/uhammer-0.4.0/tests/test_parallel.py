#! /usr/bin/env python
# encoding: utf-8
from __future__ import absolute_import, division, print_function

import os
import socket
import subprocess
import sys
import time
import warnings
from distutils.spawn import find_executable
from multiprocessing import cpu_count, current_process

import numpy as np
import pytest
from schwimmbad import MultiPool

from uhammer import (Parameters, continue_sampling, persist_every_n_iterations,
                     sample)
from uhammer._parallel import (check_if_walkers_number_of_cores_fit,
                               check_pool, core_count, runs_on_euler_node,
                               runs_on_single_euler_node)


def test_check_euler(monkeypatch):

    assert not runs_on_euler_node()

    for ip in ("10.205.0.10", "10.205.1.10", "10.205.96.0", "10.205.97.0"):

        monkeypatch.setattr(socket, "gethostbyname", lambda name: ip)
        assert runs_on_euler_node()


def test_runs_on_single_euler_node(monkeypatch):
    assert not runs_on_single_euler_node()
    monkeypatch.setenv("LSB_EFFECTIVE_RSRCREQ", "span[hosts=1]")
    assert runs_on_single_euler_node()


def test_core_count(monkeypatch):
    assert core_count() == cpu_count()
    monkeypatch.setenv("LSB_MAX_NUM_PROCESSORS", "124")
    assert core_count() == 124


def test_warning_for_oversubscription(record_warnings, regtest, monkeypatch):
    monkeypatch.setenv("OMP_NUM_THREADS", "124")
    with record_warnings() as messages:
        check_pool(MultiPool(2), 10)
        assert len(messages) == 1, messages
        print("expect warning about oversubscription:".upper(), file=regtest)
        print(messages[0], file=regtest)


def test_check_if_walkers_number_of_cores_fit(record_warnings, regtest):

    with record_warnings() as messages:
        assert not check_if_walkers_number_of_cores_fit(10, 1)
        assert len(messages) == 1, messages
        print(messages[0], file=regtest)

    with record_warnings() as messages:
        check_if_walkers_number_of_cores_fit(10, 20)
        assert len(messages) == 1, messages
        print("expect warning about less walkers than cores:".upper(), file=regtest)
        print(messages[0], file=regtest)

    print(file=regtest)

    with record_warnings() as messages:
        check_if_walkers_number_of_cores_fit(81, 20)
        assert len(messages) == 1, messages
        print(
            "expectwarning about walkers beig multiples of num workers:".upper(),
            file=regtest,
        )
        print(messages[0], file=regtest)

    print(file=regtest)

    with record_warnings() as messages:
        check_if_walkers_number_of_cores_fit(80, 20)
        assert not messages


class PoolMock(object):
    def __init__(self, size):
        self.size = size


def test_check_pool_1(record_warnings, monkeypatch, regtest):
    class PoolMock(object):
        def __init__(self, size):
            self.size = size

    with record_warnings() as messages:
        monkeypatch.setattr(socket, "gethostbyname", lambda name: "10.205.0.10")
        check_pool(PoolMock(36), 72)
        assert len(messages) == 2
        for warning in messages:
            print(warning, file=regtest)


def test_check_pool_2(record_warnings, monkeypatch, regtest):

    with record_warnings() as messages:

        # setup single node on euler:
        monkeypatch.setattr(socket, "gethostbyname", lambda name: "10.205.0.10")
        monkeypatch.setenv("LSB_EFFECTIVE_RSRCREQ", "span[hosts=1]")

        check_pool(PoolMock(36), 72)
        assert len(messages) == 1
        for warning in messages:
            print(warning, file=regtest)


def test_check_pool_3(record_warnings):

    with record_warnings() as messages:
        check_pool(PoolMock(35), 70)
        assert not messages


def lnprob(p):
    # this function must be outside the test functions to be pickable, as
    # multiprocessing pool fails for non-pickable functions.

    time.sleep(0.02)  # make it slow to increase speedup.
    return -p.a ** 2


def test_parallel_sampling(record_warnings, regtest, tmpdir):
    p = Parameters()
    p.add("a", (-1, 1))

    n_samples = 50
    n_walkers_per_parm = 14

    started = time.time()

    samples_serial, __ = sample(
        lnprob, p, [], n_walkers_per_parm, seed=42, n_samples=n_samples, parallel=False
    )

    needed_serial = time.time() - started

    started = time.time()

    with warnings.catch_warnings():
        warnings.simplefilter("ignore", UserWarning)

        samples_parallel, __ = sample(
            lnprob,
            p,
            [],
            n_walkers_per_parm,
            seed=42,
            n_samples=n_samples,
            parallel=True,
            persist=persist_every_n_iterations(1, tmpdir.join("sampler.pkl").strpath),
        )
        needed_parallel = time.time() - started

    assert np.all(samples_serial == samples_parallel), (
        samples_serial,
        samples_parallel,
    )

    # we dont check scalablity for less than four cores as my develpment machine has
    # four. For less than four cores it is enough that the assert above did not fail.
    if cpu_count() >= 4:
        assert needed_parallel < 0.6 * needed_serial, (needed_parallel, needed_serial)


def test_mpi_tasks(record_warnings, regtest, monkeypatch):

    from uhammer._parallel import mpi_tasks  # avoids UnboundLocalError

    assert mpi_tasks() is None

    if "mpi4py" in sys.modules:
        del sys.modules["mpi4py"]

    from uhammer._parallel import mpi_tasks

    assert "mpi4py" not in sys.modules, "this test is not working any more"

    class Fake(object):
        @property
        def MPI(self):
            raise ImportError("import fails on purpose.")

    try:
        sys.modules["mpi4py"] = Fake()
        # simulate mpirun:
        os.environ["MPIEXEC_FAKE"] = "LOOKSLIKEMPI"
        with record_warnings() as messages:
            assert mpi_tasks() == 1
    finally:
        del os.environ["MPIEXEC_FAKE"]
        del sys.modules["mpi4py"]

    assert len(messages) == 1, messages
    print(messages[0], file=regtest)


def test_sample_from_lambda():

    seed = 42

    p = Parameters()
    p.add("a", (-1, 1))

    lnprob = lambda s: -abs(s.a)  # noqa

    with warnings.catch_warnings():
        warnings.simplefilter("ignore", UserWarning)
        sample(
            lnprob,
            p,
            args=None,
            n_walkers_per_param=10,
            seed=seed,
            n_samples=1000,
            show_progress=False,
            parallel=True,
        )


MPI_EXE = find_executable("mpirun")
try:
    import mpipool  # noqa
except ImportError:
    HAS_MPIPOOL = False
else:
    HAS_MPIPOOL = True


@pytest.mark.skipif(HAS_MPIPOOL is False, reason="mpipool not installed")
@pytest.mark.skipif(MPI_EXE is None, reason="mpirun not found")
@pytest.mark.skipif(sys.version_info.major == 2, reason="fails with python")
def test_mpi(tmpdir):

    output_prefix = tmpdir.join("out").strpath
    output_prefix_continue = tmpdir.join("out_continue").strpath
    persist_path = tmpdir.join("sampler.pkl").strpath

    env = os.environ.copy()
    env.update({"OUTPUT_PREFIX": output_prefix, "PERSIST_PATH": persist_path})

    here = os.path.dirname(os.path.abspath(__file__))
    script = os.path.join(here, "files", "sample_line_fit_parallel.py")
    line = [MPI_EXE, "-n", "4", sys.executable, script]

    p = subprocess.Popen(
        line, stdout=subprocess.PIPE, stderr=subprocess.PIPE, env=env  # shell=True
    )
    found_error = False
    found_shape = False
    for line in iter(p.stdout.readline, b""):
        line = line.rstrip()
        print(str(line, "utf-8"))
        if line.startswith(b"++++ ERROR"):
            found_error = True
        if line.startswith(b"++++ SAMPLES"):
            found_shape = b" ".join(line.split(b" ")[2:])

    assert not found_error
    assert found_shape == b"(1500, 3)"

    pathes = set(os.path.relpath(p.strpath, tmpdir.strpath) for p in tmpdir.listdir())
    expected = [
        os.path.basename(output_prefix) + "_run_0000_worker_{}.txt".format(i)
        for i in range(1, 4)
    ] + [os.path.basename(persist_path)]

    assert pathes == set(expected)

    for worker_id, p in enumerate(expected[:-1], 1):
        full_p = tmpdir.join(p).strpath
        content = set(open(full_p))
        assert content == set((str(worker_id) + "\n",))

    samples, __ = continue_sampling(
        persist_path,
        10,
        parallel=False,
        output_prefix=output_prefix_continue,
        show_output=False,
    )
    assert samples.shape == (10, 3)
    assert os.path.exists(output_prefix_continue + "_run_0000.txt")


def test_multiprocessing(tmpdir):

    output_prefix = tmpdir.join("out").strpath
    output_prefix_continue = tmpdir.join("out_continue").strpath
    persist_path = tmpdir.join("sampler.pkl").strpath

    sigma = .5

    def gen_data():
        a0 = .5
        b0 = .5
        c0 = 1

        x = np.linspace(-2, 2, 100)
        y_measured = a0 + b0 * x + c0 * x ** 2 + sigma * np.random.randn(*x.shape)
        return x, y_measured

    p = Parameters()
    p.add("a", (0, 1))
    p.add("b", (0, 1))
    p.add("c", (0, 2))

    def lnprob(p, x, y_measured, sigma=sigma):
        print(current_process().rank)
        y = p.a + p.b * x + p.c * x ** 2
        diff = (y - y_measured) / sigma
        return -np.dot(diff.T, diff) / 2

    n_samples = 1500
    n_walkers_per_param = 200

    samples, __ = sample(
        lnprob,
        p,
        args=gen_data(),
        n_walkers_per_param=n_walkers_per_param,
        n_samples=n_samples,
        show_progress=False,
        show_output=False,
        output_prefix=output_prefix,
        parallel=True,
        verbose=True,
        persist=persist_every_n_iterations(1, persist_path),
    )

    pathes = set(os.path.relpath(p.strpath, tmpdir.strpath) for p in tmpdir.listdir())
    expected = [
        os.path.basename(output_prefix) + "_run_0000_worker_{}.txt".format(i)
        for i in range(1, cpu_count())
    ] + [os.path.basename(persist_path)]

    assert pathes == set(expected)

    for worker_id, p in enumerate(expected[:-1], 1):
        full_p = tmpdir.join(p).strpath
        content = set(open(full_p))
        assert content == set((str(worker_id) + "\n",))

    samples, __ = continue_sampling(
        persist_path,
        10,
        parallel=False,
        output_prefix=output_prefix_continue,
        show_output=False,
    )
    assert samples.shape == (10, 3)

    assert os.path.exists(output_prefix_continue + "_run_0000.txt")
