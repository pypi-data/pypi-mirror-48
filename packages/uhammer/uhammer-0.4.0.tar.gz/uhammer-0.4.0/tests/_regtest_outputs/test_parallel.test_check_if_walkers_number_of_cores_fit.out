only one worker available. switch to serial mode.
EXPECT WARNING ABOUT LESS WALKERS THAN CORES:
the number of walkers 10 is less than the number of workers 20.

EXPECTWARNING ABOUT WALKERS BEIG MULTIPLES OF NUM WORKERS:
uhammer runs with 20 workers for 81 walkers. resource usage is best if the number walkers is a multiple of number of workers.

